### PR TITLE
Part2

### PR DESCRIPTION
**Pull request by dpsproservices**
Project 4 - Part 2 / 9
 Video: Chapter 3 Part 6

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -1,48 +1,574 @@
...
+        std::cout << "warning: floating point division by zero!" << std::endl;
+        *value /= std::numeric_limits<float>::denorm_min();
+        return *this;
+    }
+    else
+    {
+        *value /= f;
+        return *this;
+    }
+}
```
Comment (matkatmusic) [inline main.cpp]: D.R.Y.  both blocks return `*this`
You changed the functionality from part1's divide function.
continue to divide by your function argument.   Warn if `f == 0.f` only.
- [ ] revise.
```
if( f == 0.f )
    issue the warning...

do the division
return *this;
```

File: main.cpp [hunk 2]
```diff
@@ -1,48 +1,574 @@
...
+        std::cerr << "warning: floating point division by zero!" << std::endl;
+        *value /= std::numeric_limits<double>::denorm_min();
+        return *this;
+    }
+    else
+    {
+        *value /= d;
+        return *this;
+    }
+}
```
Comment (matkatmusic) [inline main.cpp]: The same comment as `FloatType::divide(float)` applies here.
- [ ] revise

File: main.cpp [hunk 3]
```diff
@@ -1,48 +1,574 @@
...
+    {
+        std::cerr << "error: integer division by zero is an error and will crash the program!" << std::endl;
+        return *this;
+    } 
+    else 
+    {
+        *value /= i;
+        return *this;
+    }
+}
```
Comment (matkatmusic) [inline main.cpp]: D.R.Y. 
you're returning `*this` in both blocks.
- [ ] revise.

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 3 above)
Comment (dpsproservices) [reply main.cpp]: Ok done revised divide member functions , tested committed.

File: main.cpp [hunk 4]
```diff
@@ -1,48 +1,557 @@
...
+
+struct A {};
+struct HeapA
+{
+    A* pointerToA = nullptr;
+
+    HeapA( )
+    { 
+        pointerToA = new A;
+    }
```
Comment (matkatmusic) [inline main.cpp]: You should use the member initializer list here instead of inside the constructor body.

File: main.cpp [hunk 5]
```diff
@@ -1,48 +1,557 @@
...
+    IntType& divide(const IntType& it);
+    IntType& divide(const FloatType& ft);
+    IntType& divide(const DoubleType& dt);
+};
+
+/* FloatType member function definitions */
+FloatType::FloatType(float floatPrimitive) 
+{
+    value = new float(floatPrimitive);
+}
```
Comment (matkatmusic) [inline main.cpp]: use the member initializer list.
- [ ] revise

File: main.cpp [hunk 6]
```diff
@@ -1,48 +1,557 @@
...
+{
+    return divide( *dt.value );
+}
+
+/* DoubleType member function definitions */
+
+DoubleType::DoubleType(double doublePrimitive) 
+{
+    value = new double(doublePrimitive);
+}
```
Comment (matkatmusic) [inline main.cpp]: use the member initializer list.
- [ ] revise

File: main.cpp [hunk 7]
```diff
@@ -1,48 +1,557 @@
...
+{
+    return divide( *dt.value );
+}
+
+/* IntType member function definitions */
+
+IntType::IntType(int intPrimitive) 
+{
+    value = new int(intPrimitive);
+}
```
Comment (matkatmusic) [inline main.cpp]: use the member initializer list.
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
always prefer the member initializer list over in-constructor initialization.
always prefer in-class initialization over the member initializer list.

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**